packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSource.ts

```diff
@@ -0,0 +1,286 @@
+
+    const cachedData = privateApiRef.current.getCacheData(fetchParams);
+
+    if (cachedData != null) {
+      const rows = cachedData.rows;
+      privateApiRef.current.caches.dataSource.groupKeys = [];
+      privateApiRef.current.setRows(rows);
+      if (cachedData.rowCount) {
+        privateApiRef.current.setRowCount(cachedData.rowCount);
+      }
+      return;
+    }
+
+    const isLoading = gridRowsLoadingSelector(privateApiRef);
+    if (!isLoading) {
+      privateApiRef.current.setLoading(true);
+    }
+
+    try {
+      const getRowsResponse = await getRows(fetchParams);
+      privateApiRef.current.setCacheData(fetchParams, getRowsResponse);
+      if (getRowsResponse.rowCount) {
+        privateApiRef.current.setRowCount(getRowsResponse.rowCount);
+      }
+      privateApiRef.current.caches.dataSource.groupKeys = [];
+      privateApiRef.current.setRows(getRowsResponse.rows);
+      privateApiRef.current.setLoading(false);
+    } catch (error) {
+      privateApiRef.current.setRows([]);
+      privateApiRef.current.setLoading(false);
+      onError?.(error as Error, fetchParams);
+    }
+  }, [nestedDataManager, privateApiRef, props.unstable_dataSource?.getRows, onError]);
+
+  const queueChildrenFetch = React.useCallback(
+    (id: GridRowId) => {
+      privateApiRef.current.setChildrenLoading(id, true);
+      nestedDataManager.enqueue([id]);
+    },
+    [privateApiRef, nestedDataManager],
+  );
+
+  const fetchRowChildren = React.useCallback<GridDataSourcePrivateApi['fetchRowChildren']>(
+    async (id) => {
+      if (!props.treeData) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+      const getRows = props.unstable_dataSource?.getRows;
+      if (!getRows) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+
+      const rowNode = privateApiRef.current.getRowNode(id) as GridServerSideGroupNode;
+      if (!rowNode) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+
+      const fetchParams = { ...gridGetRowsParamsSelector(privateApiRef), groupKeys: rowNode.path };
+
+      const cachedData = privateApiRef.current.getCacheData(fetchParams);
+
+      const isLoading = gridDataSourceLoadingSelector(privateApiRef)[id] ?? false;
+      if (cachedData != null) {
+        const rows = cachedData.rows;
+        privateApiRef.current.caches.dataSource.groupKeys = rowNode.path;
+        nestedDataManager.setRequestSettled(id);
+        privateApiRef.current.updateRows(rows, false);
+        if (cachedData.rowCount) {
+          privateApiRef.current.setRowCount(cachedData.rowCount);
+        }
+        privateApiRef.current.setRowChildrenExpansion(id, true);
+        if (isLoading) {
+          privateApiRef.current.setChildrenLoading(id, false);
+        }
+        return;
+      }
+
+      if (!isLoading) {
+        privateApiRef.current.setChildrenLoading(id, true);
+      }
```

Comment:
This is needed for the groups that are auto-fetched because they are only set as `loading` when the fetch is started (when `requestStatus` becomes PENDING). As compared to when a user expands a node manually, we need to set `loading: true` right away, even if the request has to go to the QUEUED state.

Does it make sense?

Reply:
Shouldn't auto-fetched and manually triggered rows all go through the same pipeline and have the loading state set in the same place? Or is there something that differentiates them?

Reply:
So the current flow goes like this.
1. When the Grid mounts (or dependencies of the `useEffect` change), groups that are needed to be auto-fetched are picked by the `useGridDataSource` hook.
2. They are added to the `nestedDataManager` fetch queue.
3. If `maxConcurrentRequests` is less than the size of the requested rows, some of the rows go to the QUEUED state.
4. The loading icon only shows for those rows that are fetching (are in the PENDING state) right now to provide appropriate feedback to the user
5. As soon as some pending requests get SETTLED, more requests move to the PENDING state, and the loading icon shows for them

This is a kind of UX optimization.

Compared to this, when the user clicks on the expand icon, we show the loading icon right away to give quick feedback to the user even if the request has to go to the QUEUED state because of the pending requests.

Does it make more sense to set the loading icons for all the rows (in auto fetch) whether QUEUED or PENDING?

Reply:
They should both have a loading icon.